### ISSUE TITLE
displayio - pause display refresh (for setting up a bunch of items)

### ISSUE THREAD

**Issue by ladyada** [adafruit/circuitpython#1691]
the opposite of refresh_soon() :)

**Comment by deshipu** [comment]
maybe a context manager would work for this, pausing the display for the duration of executing the code inside of it

**Comment by deshipu** [comment]
something like:
```python
with display.frame: # pauses the refresh here
   ...
# unpauses and schedules a new refresh here
```

**Comment by tannewt** [comment]
I expect to experiment a bunch with frame synchronization in 4.x once we can actually update a frame faster than the display does. Thanks for making an issue for it.

**Comment by tannewt** [comment]
Ok, so I think I know what I want to replace `wait_for_frame` and `refresh_soon` with. I think we want an explicit `refresh` call that blocks until new pixels have been sent to the display (but not wait on epaper refresh.)

For fast screens it'll have a `target_fps` kwarg that takes a target frame rate and skips the pixel update if the time since the last frame is more than the gap needed for the frame rate. It'll also have an `max_delay` that will cause an exception if the time since the last updated frame is greater than the given delay.

**Comment by ladyada** [comment]
do you want to have anything for pausing so that if you have text on top of a bitmap you dont have a few characters appear at a time

**Comment by tannewt** [comment]
@ladyada You don't need to pause because a refresh will only happen when you call `refresh`. It won't auto happen at all. (though the terminal/supervisor will call it too)

**Comment by ladyada** [comment]
ohh, wait, so you will have to poke the display for each drawing? how will that work if you draw/change a bitmap like with turtle?

**Comment by tannewt** [comment]
You'll want to call refresh from time to time and the dirty area will be updated. I don't think it's too common to have code that may run much faster than the display can update.

**Comment by ladyada** [comment]
totally cool - i just wanted to know what to expect - the arduino code's like this too so i'm pretty used to it!

**Comment by deshipu** [comment]
That seems reasonable, and much less magical, I like it. I would probably rename ``max_delay`` to ``max_lag``, to make it clearer.

**Comment by tannewt** [comment]
I think we will want a `ready_for_refresh` for EInk so we can decide not to wait three minutes.

**Comment by ladyada** [comment]
yes, ready_for_refresh is a good idea :) just a note that sharp memory displays *must* be refreshed once a second, or the image disappears. so its sorta the opposite of e-ink.

**Comment by tannewt** [comment]
We're going to add `auto_refresh` as a kwarg and property to Display so that `refresh` is optional from the start.
